Cookbook/Proposals/SmarterDiffs.md

```diff
@@ -0,0 +1,97 @@
+
+#### Option 1 - assert states one by one
+
+Example:
+
+```swift
+assert: { states in
+    expect(states.first?.lastName).to(beNil())
+    expect(states.last?.lastName).to(equal("Snow"))
+}
+```
+
+This approach works fine and makes the test expectations very explicit. But it becomes overly verbose when we have to assert a lot of states one by one. It may be tempting then to make shortcuts like in the example and check for particular elements in the states array and not for all of them which makes such tests incomplete as they will not catch issues when some other state unexpectedly changed or the number of states changed, unless we explicitly test for it. Another possible issue is that if we use index subscript on state array the test run can crash if index goes out of bounds instead of failing assertion and continue running other tests.
+At the same time if a single state assert fails it's easier to understand what failed as we are comparing less values at a time.
+
+#### Option 2 - assert all states at once
+
+Example:
+
+```swift
+expect(states).to(equal([
+    makeState(status: .idle),
+    makeState(status: .loading),
+    makeState(
+        shouldShowNHSBanner: true,
+        shouldShowHealthcheckBanner: true,
+        shouldShowTestAndKitsBanner: true,
+        shouldShowGPAtHandStatusCard: true,
+        status: .loaded
+    )
+]))
+```
+
+This eliminates the issue of the first approach when some state changes may be ignored by the test. It comes with a cost of actually creating all the intermediate states that might not be directly related to what test is testing which adds some amount of boilerplate. Another issue is that when this single assert fails it's very hard to extract what actually cuased this failure from the error message until we go and check each states separately. First option also suffers from this issue when each state consists of a lot of properties but it becomes much worse with a single assert.
+
+## Motivation
+
+As demonstrated our current approaches suffer from few issues which result in not ideal developer experience when it comes to debugging test failures.
+
+## Proposed solution
+
+To improve the experience with debugging test failures we propose to use a smarter diffing approach as in [this library](https://github.com/krzysztofzablocki/Difference) by Krzysztof Zabłocki. The idea is that when comparing two instances we will be using `Mirror` type and `dump` function to first get stored properties and then get their textual representation and this way combine a in-memory snapshot of the value. This way we can recursively compare each property and only include those that differ in the final failure message. 
+
+The fact that `Mirror` only provides access to stored properties is not an issue as computed properties are products of stored properties (or in case of enums of the enum cases and thier associated values) and for same stored properties values should return the same result.
```

Comment:
```suggestion
The fact that `Mirror` only provides access to stored properties is not an issue as computed properties are products of stored properties (or in case of enums of the enum cases and their associated values) and for same stored properties values should return the same result.
```